crash when IP reservation is filtered by facility and no facility is returned

An IP reservation may not include a Facility and 
https://github.com/equinix/terraform-provider-metal/blob/main/metal/datasource_metal_precreated_ip_block.go#L91-L102 does not handle that.

```
panic: runtime error: invalid memory address or nil pointer dereference
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x18a67fb]
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0:
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: goroutine 69 [running]:
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: github.com/equinix/terraform-provider-metal/metal.dataSourceMetalPreCreatedIPBlockRead(0xc00020cd90, 0x19d8080, 0xc00013a280, 0xc00020cd90, 0x0)
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: 	github.com/equinix/terraform-provider-metal/metal/datasource_metal_precreated_ip_block.go:95 +0x39b
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).ReadDataApply(0xc000158580, 0xc00095c280, 0x19d8080, 0xc00013a280, 0xc000178798, 0x1, 0x0)
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: 	github.com/hashicorp/terraform-plugin-sdk@v1.0.0/helper/schema/resource.go:398 +0x91
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).ReadDataApply(0xc000054280, 0xc0003ad9a8, 0xc00095c280, 0xc00095c280, 0x0, 0x0)
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: 	github.com/hashicorp/terraform-plugin-sdk@v1.0.0/helper/schema/provider.go:446 +0x8f
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadDataSource(0xc00000e048, 0x1c0e290, 0xc0009b6630, 0xc000540c40, 0xc00000e048, 0xc0009b6630, 0xc0006a2a50)
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: 	github.com/hashicorp/terraform-plugin-sdk@v1.0.0/internal/helper/plugin/grpc_provider.go:1036 +0x42d
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadDataSource_Handler(0x1a74460, 0xc00000e048, 0x1c0e290, 0xc0009b6630, 0xc00019dc80, 0x0, 0x1c0e290, 0xc0009b6630, 0xc0004b2000, 0xef)
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: 	github.com/hashicorp/terraform-plugin-sdk@v1.0.0/internal/tfplugin5/tfplugin5.pb.go:3225 +0x214
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001d0160, 0x1c16478, 0xc000582a80, 0xc0000f1d00, 0xc00010f020, 0x20e3ed0, 0x0, 0x0, 0x0)
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: 	google.golang.org/grpc@v1.23.0/server.go:995 +0x482
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: google.golang.org/grpc.(*Server).handleStream(0xc0001d0160, 0x1c16478, 0xc000582a80, 0xc0000f1d00, 0x0)
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: 	google.golang.org/grpc@v1.23.0/server.go:1275 +0xd2c
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000042290, 0xc0001d0160, 0x1c16478, 0xc000582a80, 0xc0000f1d00)
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: 	google.golang.org/grpc@v1.23.0/server.go:710 +0xab
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: created by google.golang.org/grpc.(*Server).serveStreams.func1
2021-07-21T14:57:45.280-0500 [DEBUG] plugin.terraform-provider-metal_v2.1.0: 	google.golang.org/grpc@v1.23.0/server.go:708 +0xa5
2021-07-21T14:57:45.282-0500 [DEBUG] plugin: plugin process exited: path=.terraform/plugins/registry.terraform.io/equinix/metal/2.1.0/darwin_amd64/terraform-provider-metal_v2.1.0 pid=74821 error="exit status 2"
```

--- a/metal/datasource_metal_precreated_ip_block.go
+++ b/metal/datasource_metal_precreated_ip_block.go
@@ -92,6 +92,9 @@ func dataSourceMetalPreCreatedIPBlockRead(d *schema.ResourceData, meta interface
 		// lookup of not-global block
 		facility := fval.(string)
 		for _, ip := range ips {
+			if ip.Facility == nil {
+				continue
+			}
 			if ip.Public == public && ip.AddressFamily == ipv && facility == ip.Facility.Code {
 				if err := loadBlock(d, &ip); err != nil {
 					return err

--- a/metal/datasource_metal_precreated_ip_block.go
+++ b/metal/datasource_metal_precreated_ip_block.go
@@ -27,15 +27,17 @@ func dataSourceMetalPreCreatedIPBlock() *schema.Resource {
 	}
 
 	s["facility"] = &schema.Schema{
-		Type:        schema.TypeString,
-		Optional:    true,
-		Description: "Facility of the searched block. (for non-global blocks).",
+		Type:          schema.TypeString,
+		Optional:      true,
+		Description:   "Facility of the searched block. (for non-global blocks).",
+		ConflictsWith: []string{"metro"},
 	}
 
 	s["metro"] = &schema.Schema{
-		Type:        schema.TypeString,
-		Optional:    true,
-		Description: "Metro of the searched block (for non-global blocks).",
+		Type:          schema.TypeString,
+		Optional:      true,
+		Description:   "Metro of the searched block (for non-global blocks).",
+		ConflictsWith: []string{"facility"},
 	}
 
 	s["address_family"] = &schema.Schema{
@@ -84,53 +86,37 @@ func dataSourceMetalPreCreatedIPBlockRead(d *schema.ResourceData, meta interface
 		return fmt.Errorf("You can't specify facility for global IP block - addresses from global blocks can be assigned to devices across several locations")
 	}
 
-	if fok && mok {
-		return fmt.Errorf("You can't specify both facility and metro.")
-	}
-
 	if fok {
-		// lookup of not-global block
+		// lookup of block specified with facility
 		facility := fval.(string)
 		for _, ip := range ips {
 			if ip.Facility == nil {
 				continue
 			}
 			if ip.Public == public && ip.AddressFamily == ipv && facility == ip.Facility.Code {
-				if err := loadBlock(d, &ip); err != nil {
-					return err
-				}
-				break
+				return loadBlock(d, &ip)
 			}
 		}
 	} else if mok {
-		// lookup of not-global block
+		// lookup of blcok specified with metro
 		metro := mval.(string)
 		for _, ip := range ips {
 			if ip.Metro == nil {
 				continue
 			}
 			if ip.Public == public && ip.AddressFamily == ipv && metro == ip.Metro.Code {
-				if err := loadBlock(d, &ip); err != nil {
-					return err
-				}
-				break
+				return loadBlock(d, &ip)
 			}
 		}
 	} else {
-		// lookup of global block
+		// lookup of blocks not specified with facility or metro
 		for _, ip := range ips {
-			if ip.Public == public && ip.AddressFamily == ipv && ip.Global {
-				if err := loadBlock(d, &ip); err != nil {
-					return err
-				}
-				break
+			if ip.Public == public && ip.AddressFamily == ipv && global == ip.Global {
+				return loadBlock(d, &ip)
 			}
 		}
 
 	}
-	if d.Get("cidr_notation") == "" {
-		return fmt.Errorf("Could not find matching reserved block, all IPs were %v", ips)
-	}
-	return nil
+	return fmt.Errorf("Could not find matching reserved block, all IPs were %v", ips)
 
 }

--- a/metal/datasource_metal_precreated_ip_block_test.go
+++ b/metal/datasource_metal_precreated_ip_block_test.go
@@ -52,6 +52,7 @@ resource "metal_device" "test" {
 }
 
 data "metal_precreated_ip_block" "test" {
+    facility         = metal_device.test.deployed_facility
     project_id       = metal_device.test.project_id
     address_family   = 6
     public           = true

--- a/metal/datasource_metal_precreated_ip_block_test.go
+++ b/metal/datasource_metal_precreated_ip_block_test.go
@@ -52,7 +52,7 @@ resource "metal_device" "test" {
 }
 
 data "metal_precreated_ip_block" "test" {
-    facility         = "ewr1"
+    metro            = "ny"
     project_id       = metal_device.test.project_id
     address_family   = 6
     public           = true

--- a/metal/datasource_metal_precreated_ip_block_test.go
+++ b/metal/datasource_metal_precreated_ip_block_test.go
@@ -52,7 +52,6 @@ resource "metal_device" "test" {
 }
 
 data "metal_precreated_ip_block" "test" {
-    metro            = "ny"
     project_id       = metal_device.test.project_id
     address_family   = 6
     public           = true

--- a/metal/datasource_metal_precreated_ip_block_test.go
+++ b/metal/datasource_metal_precreated_ip_block_test.go
@@ -8,7 +8,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 )
 
-func TestAccMetalPreCreatedIPBlock_Basic(t *testing.T) {
+func TestAccMetalDatasourcePreCreatedIPBlock_Basic(t *testing.T) {
 
 	rs := acctest.RandString(10)
 
@@ -17,7 +17,7 @@ func TestAccMetalPreCreatedIPBlock_Basic(t *testing.T) {
 		Providers: testAccProviders,
 		Steps: []resource.TestStep{
 			{
-				Config: testPreCreatedIPBlockConfig_Basic(rs),
+				Config: testDatasourcePreCreatedIPBlockConfig_Basic(rs),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttrSet(
 						"data.metal_precreated_ip_block.test", "cidr_notation"),
@@ -35,7 +35,7 @@ func TestAccMetalPreCreatedIPBlock_Basic(t *testing.T) {
 	})
 }
 
-func testPreCreatedIPBlockConfig_Basic(name string) string {
+func testDatasourcePreCreatedIPBlockConfig_Basic(name string) string {
 	return fmt.Sprintf(`
 
 resource "metal_project" "test" {
@@ -44,23 +44,23 @@ resource "metal_project" "test" {
 
 resource "metal_device" "test" {
   hostname         = "tfacc-test-device-ip-blockt"
-  plan             = "t1.small.x86"
-  facilities       = ["ewr1"]
+  plan             = "c3.small.x86"
+  metro            = "ny"
   operating_system = "ubuntu_16_04"
   billing_cycle    = "hourly"
-  project_id       = "${metal_project.test.id}"
+  project_id       = metal_project.test.id
 }
 
 data "metal_precreated_ip_block" "test" {
     facility         = "ewr1"
-    project_id       = "${metal_device.test.project_id}"
+    project_id       = metal_device.test.project_id
     address_family   = 6
     public           = true
 }
 
 resource "metal_ip_attachment" "test" {
-    device_id = "${metal_device.test.id}"
-    cidr_notation = "${cidrsubnet(data.metal_precreated_ip_block.test.cidr_notation,8,2)}"
+    device_id = metal_device.test.id
+    cidr_notation = cidrsubnet(data.metal_precreated_ip_block.test.cidr_notation,8,2)
 }
 `, name)
 }

